tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuning.scala

```diff
@@ -0,0 +1,616 @@
+        recommendations(key) = recommendationVal
+      }
+    }
+  }
+
+  def appendRecommendation(key: String, value: String): Unit = {
+    val recomRecord = recommendations.getOrElseUpdate(key,
+      new RecommendationEntry(key, null, null))
+    recomRecord.setRecommendedValue(value)
+    if (recomRecord.originalValue == null && value != null) {
+      // add a comment
+      appendComment(s"'$key' was not set.")
+    }
+  }
+
+  def appendRecommendationForMemoryMB(key: String, value: String): Unit = {
+    if (value != null) {
+      appendRecommendation(key, s"${value}m")
+    }
+  }
+
+  /**
+   * Recommendation for 'spark.executor.instances' based on number of gpus and workers.
+   * Assumption - In case GPU properties are not available, it assumes 1 GPU per worker.
+   */
+  def calcExecInstances(): Int = {
+    // spark.executor.instances
+    clusterProps.gpu.getCount * clusterProps.system.numWorkers
+  }
+
+  /**
+   * Recommendation for 'spark.executor.cores' based on number of cpu cores and gpus.
+   */
+  def calcNumExecutorCores(): Int = {
+    // spark.executor.cores
+    val executorsPerNode = clusterProps.gpu.getCount
+    if (executorsPerNode == 0) {
+      1
+    } else {
+      Math.max(1, clusterProps.system.numCores / executorsPerNode)
+    }
+  }
+
+  /**
+   * Recommendation for 'spark.task.resource.gpu.amount' based on num of cpu cores.
+   */
+  def calcTaskGPUAmount(numExecCoresClaculator: () => Int): Double = {
+    // spark.task.resource.gpu.amount
+    val numExecutorCores =  numExecCoresClaculator.apply()
+    if (numExecutorCores == 0) {
+      0.0
+    } else {
+      1.0 / numExecutorCores
+    }
+  }
+
+  /**
+   * Recommendation for 'spark.rapids.sql.concurrentGpuTasks' based on gpu memory.
+   */
+  def calcGpuConcTasks(): Long = {
+    // spark.rapids.sql.concurrentGpuTasks
+    Math.min(MAX_CONC_GPU_TASKS,
+      convertToMB(clusterProps.gpu.memory) / DEF_GPU_MEM_PER_TASK_MB)
+  }
+
+  private def calcExecutorContainerMem(): Double = {
+    // account for system overhead
+    val usableWorkerMem =
+      Math.max(0, convertToMB(clusterProps.system.memory) - DEF_SYSTEM_RESERVE_MB)
+    val executorsPerNode = clusterProps.gpu.getCount
+    if (executorsPerNode == 0) {
+      0.0
+    } else {
+      (1.0 * usableWorkerMem) / executorsPerNode
+    }
+  }
+
+  /**
+   * Recommendation for 'spark.executor.memory' based on system memory, cluster scheduler
+   * and num of gpus.
+   */
+  def calcExecutorMemory(executorContainerMemCalculator: () => Double,
+      numExecCoresClaculator: () => Int): Long = {
+    // spark.executor.memory
+    // reserve 10% of heap as memory overhead
+    val maxExecutorHeap = Math.max(0,
+      executorContainerMemCalculator.apply() * (1 - DEF_HEAP_OVERHEAD_FRACTION)).toInt
+    // give up to 2GB of heap to each executor core
+    Math.min(maxExecutorHeap, DEF_HEAP_PER_CORE_MB * numExecCoresClaculator.apply())
```

Comment:
I think maxExecutorHeap can be 0 here.  ie executorContainerMemCalculator  => calcExecutorContainerMem() returns 0 and max above is 0 so min here chooses 0.